### ISSUE TITLE
"Pattern synonym with no type signature"

### ISSUE THREAD

**Issue by aryairani** [haskell/haskell-language-server#1205]
The Problems panel shows me a million "Pattern synonym with no type signature" warnings for my project, even though I don't care about those, they are disabled in my build settings, and don't appear with `stack build`. Is there a way to disable these warnings, like we are able to do for hlint suggestions?

Thanks!

**Comment by ndmitchell** [comment]
Thanks for the report. I've transferred it to the HLS repo, as the issues you are seeing ultimately come from that. Is your project public?

**Comment by aryairani** [comment]
Thanks — it is, yes: https://github.com/unisonweb/unison

**Comment by sheaf** [comment]
Same thing on my end, I have some modules that include pattern synonyms which lack a type signature. These modules all have the PRAGMA

```haskell
{-# OPTIONS_GHC -fno-warn-missing-pattern-synonym-signatures #-}
```

yet I'm still getting the warnings in haskell-language-server in VS Code.

See for instance [this module](https://gitlab.com/sheaf/fir/-/blob/master/src/FIR/AST/Prim.hs).

**Comment by mrBliss** [comment]
I looked into this.

**TL;DR:** I believe this is due to GHC reporting the incorrect `Reason` flag for a warning.

In https://github.com/haskell/ghcide/pull/232, `enableTopLevelWarnings` was introduced, which gets called in [`typecheckModule`](https://github.com/haskell/haskell-language-server/blob/a44f4b41890745ce8d8f38f577f5f6641ef2ed94/ghcide/src/Development/IDE/Core/Compile.hs#L129):

https://github.com/haskell/haskell-language-server/blob/a44f4b41890745ce8d8f38f577f5f6641ef2ed94/ghcide/src/Development/IDE/Core/Compile.hs#L360-L366

While this may seem suspicious at first, it is fine, because the enabled warnings are hidden by `hideDiag`: https://github.com/haskell/haskell-language-server/blob/a44f4b41890745ce8d8f38f577f5f6641ef2ed94/ghcide/src/Development/IDE/Core/Compile.hs#L387-L391

So whenever a warning was produced because of `warning` (e.g., `Opt_WarnMissingPatternSynonymSignatures` or `Opt_WarnMissingPatternSynonymSignatures`) while the corresponding flag was *not* enabled for the module (it must be that `enableTopLevelWarnings` enabled it), the warning is hidden.

Unfortunately, I believe GHC is giving us the wrong reason when producing a warning for a missing pattern synonym signature. See the debug output below. I believe the `Reason` is coming straight from GHC, `ghcide` is not interfering.

```haskell
( Reason Opt_WarnMissingSignatures
, ( NormalizedFilePath "<file>"
  , ShowDiag
  , Diagnostic {
          _range = Range {
          _start = Position {_line = 8, _character = 0}
        , _end = Position {_line = 8, _character = 24}
      }
      , _severity = Just DsWarning
      , _code = Nothing
      , _source = Just "typecheck"
      , _message = "Pattern synonym with no type signature:\n  pattern I :: forall a. a -> Identity a"
      , _tags = Nothing
      , _relatedInformation = Nothing
    }
  )
)
( Reason Opt_WarnMissingSignatures
, ( NormalizedFilePath "<file>"
  , ShowDiag
  , Diagnostic {
          _range = Range {
          _start = Position {_line = 12, _character = 0}
        , _end = Position {_line = 12, _character = 3}
      }
      , _severity = Just DsWarning
      , _code = Nothing
      , _source = Just "typecheck"
      , _message = "Top-level binding with no type signature: foo :: Bool"
      , _tags = Nothing
      , _relatedInformation = Nothing
    }
  )
)
```
Even though the first warning was for a missing pattern synonym signature, the reason was `Opt_WarnMissingSignatures` instead of `Opt_WarnMissingPatternSynonymSignatures`.

I believe this is because in GHC (8.10.3, same for HEAD) when both `Opt_WarnMissingSignatures` and `Opt_WarnMissingPatternSynonymSignatures` are enabled, the first one takes precedence and will be used as the reason for [creating the warning](https://gitlab.haskell.org/ghc/ghc/-/blob/ghc-8.10.3-release/compiler/rename/RnNames.hs#L1308-1309). The logic in [`add_sig_warns`](https://gitlab.haskell.org/ghc/ghc/-/blob/ghc-8.10.3-release/compiler/rename/RnNames.hs#L1301-1305) and `add_warngs` seems wrong to me: there are two parallel pattern matches going on on the three `warn_*` bools, resulting in the inconsistency. I'm not sure whether this is intentional or not.

**Comment by mrBliss** [comment]
> I looked into this.
> 
> **TL;DR:** I believe this is due to GHC reporting the incorrect `Reason` flag for a warning.
> 
> [..]

I opened a MR for GHC that should fix this: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4869

**Comment by michaelpj** [comment]
Fixed upstream!
